sherpa/data.py

```diff
@@ -381,6 +399,11 @@ def mask(self):
     def mask(self, val):
         if (val is True) or (val is False):
             self._mask = val
+        # if val is of type np.bool_ and True, it was failed the previous test because
+        # "is True" compares with Python "True" singelton.
+        # Yet, we do not want to allow arbitrary values that evaluate as True.
+        elif (val is numpy.ma.nomask) or (numpy.isscalar(val) and val and isinstance(val, numpy.bool_)):
+            self._mask = True
```

Comment:
I would have expected something like

```
elif val is numpy.ma.mask:
    salf._mask = True
elif numpy.isscalar(val) and isinstance(val, numy.bool_):
    self._mask = val  # do we need to convert from numpy bool to bool?
elif ...
```

because don't we want `_mask` to be False if `val` is `numpy.bool_(0)`? Or does that get handled by another check in the chain?

Reply:
I guess the next question is whether we have any tests of this to check our logic makes sense? I know our existing test set doesn't do a good job of exploring the edge cases, so it would be interesting to know if there's any conditions that aren't evaluated.

Reply:
Well, I've added a LOT of tests (about a dozen, but they are all parameterized, (e.g. with and without mask) so I've lost track of the exact number, but I believe it's > 50.

Reply:
I meant have you checked the HTML coverage output to see if anything isn't being covered? It does require running the tests with coverage turned on of course, but I think it should just be 'pytest --cov=sherpa` and then `coverage html` (I never remember the option to name the output directory, or indeed the exact commands you need to use).

Reply:
I've rebased the branch onto master (so I can build it with a half-way recent set of software) and the coverage for the following (with XSPEC/astropy/matplotlib installed and the test dataset) indicates that lines 415 and 419 were never True, so their code was never executed. Which is why there's fancy unicode in the output below

```
409    def mask(self, val): 
410        if (val is True) or (val is False): 
411            self._mask = val 
412        # if val is of type np.bool_ and True, it failed the previous test because 
413        # "is True" compares with Python "True" singelton. 
414        # Yet, we do not want to allow arbitrary values that evaluate as True. 
415        elif val is numpy.ma.nomask: 415 ↛ 416
416            self._mask = True 
417        elif numpy.isscalar(val) and isinstance(val, numpy.bool_): 
418            self._mask = bool(val) 
419        elif (val is None) or numpy.isscalar(val): 419 ↛ 420
420            raise DataErr('ismask') 
421        else: 
422            self._mask = numpy.asarray(val, numpy.bool_) 
```

Reply:
I think we want to make sure to have a test of 'val is numpy.ma.nomask' evaluating to True.

Reply:
This line is only triggered when directly setting the mask (when the mask is inferred from the mask attribute of an array on initialization, then it's already handelded i nthe `_check_y` function and never gets to this point. Added test.

Reply:
Thanks,